src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -401,69 +509,45 @@ final private SecretKey generateSecretKey() throws NoSuchAlgorithmException {
         keygen.init(KEY_SIZE, mRandom);
         return keygen.generateKey();
     }
-
-    /**
-     * Supported API >= 18 PrivateKey is stored in AndroidKeyStore. Loads key
-     * from the file if it exists. If not exist, it will generate one.
-     * 
-     * @return
-     * @throws IOException
-     * @throws GeneralSecurityException
-     */
+    
     @TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
-    final synchronized private SecretKey getSecretKeyFromAndroidKeyStore() throws IOException,
-            GeneralSecurityException {
-
-        // Loading file and unwrapping this key is causing performance issue.
-        if (sSecretKeyFromAndroidKeyStore != null) {
-            return sSecretKeyFromAndroidKeyStore;
-        }
-
-        // Store secret key in a file after wrapping
-        File keyFile = new File(mContext.getDir(mContext.getPackageName(), Context.MODE_PRIVATE),
-                ADALKS);
-
-        if (mKeyPair == null) {
-            mKeyPair = getKeyPairFromAndroidKeyStore();
-            Logger.v(TAG, "Retrived keypair from androidKeyStore");
-        }
-
-        Cipher wrapCipher = Cipher.getInstance(WRAP_ALGORITHM);
-        // If keyfile does not exist, it needs to generate one
+    private synchronized SecretKey getUnwrappedSecretKey(final File keyFile) throws GeneralSecurityException, IOException {
+        Logger.v(TAG, "Reading SecretKey");
+        SecretKey unwrappedSecretKey = null;
+        
         if (!keyFile.exists()) {
-            Logger.v(TAG, "Key file does not exists");
-            final SecretKey key = generateSecretKey();
-            Logger.v(TAG, "Wrapping SecretKey");
-            final byte[] keyWrapped = wrap(wrapCipher, key);
-            Logger.v(TAG, "Writing SecretKey");
-            writeKeyData(keyFile, keyWrapped);
-            Logger.v(TAG, "Finished writing SecretKey");
+            throw new IOException("Key file to read does not exist");
         }
-
-        // Read from file again
-        Logger.v(TAG, "Reading SecretKey");
+        
         try {
-            final byte[] encryptedKey = readKeyData(keyFile);
-            if (encryptedKey == null || encryptedKey.length == 0) {
+            final byte[] wrappedSecretKey = readKeyData(keyFile);
+            if (wrappedSecretKey == null || wrappedSecretKey.length == 0) {
                 throw new UnrecoverableKeyException("Couldn't find encrypted key in file");
             }
             
-            sSecretKeyFromAndroidKeyStore = unwrap(wrapCipher, encryptedKey);
+            final PrivateKey privateKey = mKeyPair.getPrivate();
+            if (privateKey == null) {
+                throw new UnrecoverableKeyException("Retrieved private key is empty.");
```

Comment:
I want to know how this exception is surfaced to the developer. I wish we handle this internally (may be that's what is already happening).